src/platform/silabs/tracing/SilabsPowerTracing.cpp

```diff
@@ -0,0 +1,184 @@
+#include "SilabsPowerTracing.h"
+#include <lib/support/CodeUtils.h>
+#include <platform/silabs/tracing/SilabsTracingConfig.h>
+
+// Need to use the sleeptimer for power tracing. The RAIL timer doesn't tick in EM2 and creates invalid timestamps for this
+// application.
+#ifndef SILABS_GET_SLEEPTIMER_TIME
+#if defined(SL_RAIL_LIB_MULTIPROTOCOL_SUPPORT) && SL_RAIL_LIB_MULTIPROTOCOL_SUPPORT
+#include "sl_sleeptimer.h"
+#define SILABS_GET_SLEEPTIMER_TIME() uint32_t((sl_sleeptimer_get_tick_count64() * 1000ULL) / sl_sleeptimer_get_timer_frequency())
+#else
+// For unit tests, this should be defined by the test file before including the implementation
+// If not defined, provide a minimal default
+#define SILABS_GET_SLEEPTIMER_TIME() 0
+#endif
+#endif // SILABS_GET_SLEEPTIMER_TIME
+
+namespace chip {
+namespace Tracing {
+namespace Silabs {
+
+// Timer callback function - defined before use
+namespace {
+void OnPowerManagerStatisticsTimer(void * argument)
+{
+    SilabsPowerTracing::Instance().OutputPowerManagerTraces();
+}
+} // namespace
+
+// Define the static singleton instance
+SilabsPowerTracing SilabsPowerTracing::sInstance;
+
+// singleton instance constructor
+SilabsPowerTracing::SilabsPowerTracing() :
+    mEnergyTraces(nullptr), mEnergyTraceCount(0), mPowerManagerEmTransitionEventHandle(),
+    mPowerManagerEmTransitionEventInfo{ (SL_POWER_MANAGER_EVENT_TRANSITION_ENTERING_EM0 |
+                                         SL_POWER_MANAGER_EVENT_TRANSITION_ENTERING_EM1 |
+                                         SL_POWER_MANAGER_EVENT_TRANSITION_ENTERING_EM2),
+                                        StaticPowerManagerTransitionCallback },
+    mStatisticsTimer(nullptr), mInitialized(false)
+{}
+
+CHIP_ERROR SilabsPowerTracing::Init()
+{
+    CHIP_ERROR err              = CHIP_NO_ERROR;
+    bool powerManagerSubscribed = false;
+
+    // Return early if already initialized
+    if (mInitialized)
+    {
+        return CHIP_NO_ERROR;
+    }
+
+    // Allocate energy trace storage
+    if (mEnergyTraces == nullptr)
+    {
+        mEnergyTraces = static_cast<EnergyTrace *>(calloc(SILABS_TRACING_ENERGY_TRACES_MAX, sizeof(EnergyTrace)));
+        if (mEnergyTraces == nullptr)
+        {
+            err = CHIP_ERROR_NO_MEMORY;
+        }
+    }
+
+    if (err == CHIP_NO_ERROR)
```

Comment:
I don't think the if has any value here. if something failed before we would return early.